File: java/src/main/java/com/pubnub/api/endpoints/remoteaction/RemoteAction.java
package com.pubnub.api.endpoints.remoteaction;

import com.pubnub.api.PubNubException;
import com.pubnub.api.callbacks.PNCallback;

public interface RemoteAction<Output> {
    Output sync() throws PubNubException;

    void async(PNCallback<Output> callback);

    void retry();

    void silentCancel();
}

File: kotlin/src/main/kotlin/com/pubnub/api/endpoints/remoteaction/RemoteAction.kt
package com.pubnub.api.endpoints.remoteaction

import com.pubnub.api.PubNubException
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.models.consumer.PNStatus
import com.pubnub.core.CoreRemoteAction

interface ExtendedRemoteAction<Output> : RemoteAction<Output> {
    fun operationType(): PNOperationType
    fun retry()
}

interface RemoteAction<Output> : Cancelable, CoreRemoteAction<Output, PNStatus> {
    @Throws(PubNubException::class)
    fun sync(): Output?
}

interface Cancelable {
    fun silentCancel()
}
